src/TransferInternal.cs

```diff
@@ -204,26 +228,54 @@ public void UpdateProgress(long bytesTransferred)
         {
             BytesTransferred = bytesTransferred;
 
-            var ts = DateTime.UtcNow - (lastProgressTime ?? (StartTime ?? DateTime.UtcNow));
-
-            // only recompute the average speed if enough time has passed since the last computation, and if the transfer is still
-            // in progress. use a moving average to account for ramp-up and variation.
-            if (ts.TotalMilliseconds >= progressUpdateLimit && !State.HasFlag(TransferStates.Completed))
+            // that's odd! the transfer hasn't transitioned into InProgress yet. just ignore it..
+            if (!StartTime.HasValue)
             {
-                var currentSpeed = (BytesTransferred - lastProgressBytes) / (ts.TotalMilliseconds / 1000d);
-                AverageSpeed = !speedInitialized ? currentSpeed : ((currentSpeed - AverageSpeed) * speedAlpha) + AverageSpeed;
-                speedInitialized = true;
-                lastProgressTime = DateTime.UtcNow;
-                lastProgressBytes = BytesTransferred;
+                return;
             }
 
-            // once the transfer is complete, compute the actual average speed from total duration and transfer size
+            // if the state is Completed, we're guaranteed to have both StartTime and EndTime
+            // it's possible that StartTime = EndTime, and if that's the case, substitute 1ms for the duration
             if (State.HasFlag(TransferStates.Completed))
             {
-                var start = StartTime ?? EndTime.Value.AddMilliseconds(-1);
-                var duration = (EndTime.Value - start).TotalMilliseconds / 1000d;
+                var duration = Math.Max(1, (EndTime.Value - StartTime.Value).TotalMilliseconds) / 1000d;
+                var totalSpeed = (BytesTransferred - StartOffset) / duration;
+                AverageSpeed = totalSpeed;
+
+                return;
+            }
+
+            // if we've transferred all of the data but not yet transitioned into Completed, we won't have an EndTime
+            // yet, so we'll have to use the current time; the transition to Completed will happen soon!
+            if (Size.HasValue && BytesTransferred >= Size.Value)
```

Comment:
this fixes a bug that was causing transfers to report zero speed if they completed in under 1 second